lib/parser/models.ex

```diff
@@ -0,0 +1,451 @@
+      import Thrift.Parser.Conversions
+
+      @spec new(char_list) :: %Include{}
+      def new(path) do
+        %Include{path: List.to_string(path)}
+      end
+    end
+
+    defmodule Constant do
+      @moduledoc """
+      A Thrift constant.
+      Constants of any primitive or container type can be created in Thrift.
+      """
+      defstruct name: nil, value: nil, type: nil
+      import Thrift.Parser.Conversions
+
+      @spec new(char_list, Literals.t, Types.t) :: %Constant{}
+      def new(name, val, type) do
+        %Constant{name: atomify(name), value: cast(type, val), type: type}
+      end
+    end
+
+    defmodule TEnum do
+      @moduledoc """
+      A Thrift enumeration
+      An enumeration contains names and (usually sequential) values, and
+      allows you to map from one to the other.
+      """
+      @type enum_value :: bitstring | integer
+
+      defstruct name: nil, values: []
+      import Thrift.Parser.Conversions
+
+      @spec new(char_list, %{char_list => enum_value}) :: %TEnum{}
+      def new(name, values) do
+        values = values
+        |> Enum.with_index
+        |> Enum.map(fn
+          {{name, value}, _index} ->
+            {atomify(name),  value}
+
+          {name, index} ->
+            {atomify(name), index + 1}
+        end)
+
+        %TEnum{name: atomify(name), values: values}
+      end
+    end
+
+    defmodule Field do
+      @moduledoc """
+      A Thrift field.
+
+      Fields define a named type and can occur in functions, structs, unions,
+      exceptions and the parameter list and `throws` clauses of functions.
+
+      Fields can refer to each other. These are represented by the FieldReference
+      type.
+
+      This module also contains some utilities for validating and fixing up fields.
+      """
+      @type printable :: String.t | atom
+
+      defstruct id: nil, name: nil, type: nil, required: :default, default: nil
+      import Thrift.Parser.Conversions
+
+      @spec new(integer, boolean, TTypes.thrift_type, char_list, TTypes.literal) :: %Field{}
+      def new(id, required, type, name, default) do
+
+        casted_default = cast(type, default)
+
+        %Field{id: id,
+               type: type,
+               name: atomify(name),
+               required: required,
+               default: casted_default}
+      end
+
+      @spec build_field_list(printable, [%Field{}]) :: [%Field{}]
+      def build_field_list(parent_name, fields) do
+        fields
+        |> update_ids(parent_name)
+        |> validate_ids(parent_name)
+      end
+
+      defp validate_ids(fields, name) do
+        dupes = fields
+        |> Enum.group_by(&(&1.id))
+        |> Enum.filter(fn {_, v} -> length(v) > 1 end)
+
+        unless Enum.empty?(dupes) do
+          {id, fields} = List.first(dupes)
+
+          names = fields
+          |> Enum.map(&("#{name}.#{&1.name}"))
+          |> Enum.join(", ")
+
+          raise "Error: #{names} share field number #{id}."
```

Comment:
This prints too with the formatting we want. It also stops compilation, which I think is what we want to do.